src/typechecker.ts

```diff
@@ -0,0 +1,788 @@
+        break;
+      }
+      return typeFromExprType(biblioEntry.signature.return);
+    }
+    case 'tilde': {
+      if (expr.contents.length === 1 && expr.contents[0].name === 'text') {
+        return { kind: 'enum value', value: expr.contents[0].contents };
+      }
+      break;
+    }
+    case 'star': {
+      if (expr.contents.length === 1 && expr.contents[0].name === 'text') {
+        const text = expr.contents[0].contents;
+        if (text === 'null') {
+          return { kind: 'null' };
+        } else if (text === 'undefined') {
+          return { kind: 'undefined' };
+        } else if (text === 'true') {
+          return { kind: 'concrete boolean', value: true };
+        } else if (text === 'false') {
+          return { kind: 'concrete boolean', value: false };
+        } else if (text.startsWith('"') && text.endsWith('"')) {
+          return { kind: 'concrete string', value: text };
+        }
+      }
+
+      break;
+    }
+  }
+  return { kind: 'unknown' };
+}
+
+function typeFromExprType(type: BiblioType): Type {
+  switch (type.kind) {
+    case 'union': {
+      return type.types.map(typeFromExprType).reduce(join);
+    }
+    case 'list': {
+      return {
+        kind: 'list',
+        of: type.elements == null ? { kind: 'unknown' } : typeFromExprType(type.elements),
+      };
+    }
+    case 'completion': {
+      if (type.completionType === 'abrupt') {
+        return { kind: 'completion', of: { kind: 'never' } };
+      }
+      return {
+        kind: 'completion',
+        of:
+          type.typeOfValueIfNormal == null
+            ? { kind: 'unknown' }
+            : typeFromExprType(type.typeOfValueIfNormal),
+      };
+    }
+    case 'opaque': {
+      const text = type.type;
+      if (text.startsWith('"') && text.endsWith('"')) {
+        return { kind: 'concrete string', value: text.slice(1, -1) };
+      }
+      if (text.startsWith('~') && text.endsWith('~')) {
+        return { kind: 'enum value', value: text.slice(1, -1) };
+      }
+      if (text === 'a string' || text === 'strings') {
+        return { kind: 'string' };
+      }
+      if (text === 'a Number' || text === 'Numbers') {
+        return { kind: 'number' };
+      }
+      if (text === 'a Boolean' || text === 'Booleans') {
+        return { kind: 'boolean' };
+      }
+      if (text === 'an integral Number' || text === 'integral Numbers') {
+        return { kind: 'integral number' };
+      }
+      if (text === 'a mathematical value' || text === 'mathematical values') {
+        return { kind: 'real' };
+      }
+      if (text === 'an integer' || text === 'integers') {
+        return { kind: 'integer' };
+      }
+      if (text === 'a non-negative integer' || text === 'non-negative integers') {
+        return { kind: 'non-negative integer' };
+      }
+      if (text === '*null*') {
+        return { kind: 'null' };
+      }
+      if (text === '*undefined*') {
+        return { kind: 'undefined' };
+      }
+      break;
```

Comment:
Instead of returning `{ kind: 'unknown' }` here, why not have an `opaque` kind that just compares textually? Like if I have an AO that returns "a Foo" and another AO that accepts "a Bar", that should be rejected.

Reply:
Specifically because that should _not_ be rejected. Maybe all Foos are Bars. We don't have enough information to say definitively that that's wrong.

Reply:
Okay. Let's make sure there's a positive test for it then.

Reply:
We could also probably use a mode that operates like this to find other useful low-hanging fruit we could support.

Reply:
Also, `a BigInt` should be the type for bigints, right?

Reply:
Those are _parameter_ types, rather than expressions; there's a different path for converting types in signatures into type-system types than there is for converting expressions in algorithms into type-system types.